sql/sys_vars.ic

```diff
@@ -2623,37 +2623,40 @@ public:
   virtual bool do_check(THD *thd, set_var *var)
   {
     if (!Sys_var_enum::do_check(thd, var))
+    {
+      var->save_result.time.time_type= MYSQL_TIMESTAMP_NONE;
       return false;
-    MYSQL_TIME ltime;
-    bool res= var->value->get_date(&ltime, 0);
-    if (!res)
+    }
+    bool res= var->value->get_date(&var->save_result.time, 0);
+    if (res)
     {
-      var->save_result.ulonglong_value= SYSTEM_TIME_AS_OF;
+      var->save_result.time.time_type= MYSQL_TIMESTAMP_ERROR;
     }
     return res;
   }
 
 private:
   bool update(set_var *var, vers_asof_timestamp_t &out, system_variables& pool)
   {
+    MYSQL_TIME &ltime= var->save_result.time;
     bool res= false;
-    out.type= static_cast<enum_var_type>(var->save_result.ulonglong_value);
-    if (out.type == SYSTEM_TIME_AS_OF)
+
+    if (!var->value || ltime.time_type == MYSQL_TIMESTAMP_NONE)
     {
-      if (var->value)
-      {
-        MYSQL_TIME ltime;
-        uint error;
-        res= var->value->get_date(&ltime, 0);
-        out.unix_time= pool.time_zone->TIME_to_gmt_sec(&ltime, &error);
-        out.second_part= ltime.second_part;
-        res|= (error != 0);
-      }
-      else // set DEFAULT from global var
-      {
-        out= global_var(vers_asof_timestamp_t);
-        res= false;
-      }
+      out.type= SYSTEM_TIME_UNSPECIFIED;
+    }
+
+    if (var->value && ltime.time_type >= 0) // any valid value is ok
+    {
+      uint error;
+      out.type       = SYSTEM_TIME_AS_OF;
+      out.unix_time  = pool.time_zone->TIME_to_gmt_sec(&ltime, &error);
+      out.second_part= ltime.second_part;
+      res= (error != 0);
+    }
+    else // set DEFAULT from global var
+    {
+      out= global_var(vers_asof_timestamp_t);
```

Comment:
DEFAULT is SYSTEM_TIME_UNSPECIFIED.

Reply:
Yes, right. This line is the case for session var.
For global var it works with combination with another `if` several lines above:
```C
    if (!var->value || ltime.time_type == MYSQL_TIMESTAMP_NONE)
    {
      out.type= SYSTEM_TIME_UNSPECIFIED;
    }
```

So, if `default` is set to session var, it copies the value from global var;
For global var, it does `global_var = global_var`